inst/examples/arrays.cpp

```diff
@@ -5,63 +5,75 @@ class arrays {
   typedef double real_t;
   typedef dust::no_data data_t;
 
-  struct init_t {
+  struct shared_t {
     int dim_r;
     int dim_x;
-    std::array<real_t, 3> initial_x;
-    real_t initial_y;
     int n;
+    real_t initial_y;
+    std::array<real_t, 3> initial_x;
+  };
+
+  struct internal_t {
     std::array<real_t, 3> r;
   };
-  arrays(const init_t& pars): internal(pars) {
+
+  arrays(const dust::pars_t<arrays>& pars) :
+    shared(pars.shared), internal(pars.internal) {
   }
+
   size_t size() {
-    return 1 + internal.dim_x;
+    return 1 + shared->dim_x;
   }
+
   std::vector<real_t> initial(size_t step) {
-    std::vector<real_t> state(1 + internal.dim_x);
-    state[0] = internal.initial_y;
-    std::copy(internal.initial_x.begin(), internal.initial_x.end(),
+    std::vector<real_t> state(1 + shared->dim_x);
+    state[0] = shared->initial_y;
+    std::copy(shared->initial_x.begin(),
+              shared->initial_x.end(),
               state.begin() + 1);
     return state;
   }
-  #ifdef __NVCC__
-  __device__
-  #endif
-  void update(size_t step, const real_t * state, dust::rng_state_t<real_t>& rng_state, real_t * state_next) {
+
+  void update(size_t step, const real_t * state,
+              dust::rng_state_t<real_t>& rng_state, real_t * state_next) {
     const real_t * x = state + 1;
     const real_t y = state[0];
-    state_next[0] = y + internal.n;
-    for (int i = 0; i < internal.dim_x; ++i) {
+    state_next[0] = y + shared->n;
+    for (int i = 0; i < shared->dim_x; ++i) {
       state_next[i] = x[i] + internal.r[i];
     }
   }
 
 private:
-  init_t internal;
+  std::shared_ptr<const shared_t> shared;
```

Comment:
Why is this one `const`, but others aren't?

Reply:
Due to `using` in `dust.hpp` `dust::shared_ptr<T>` is the same thing as `std::shared_ptr<const typename typedef T::shared_t>` (or here `std::shared_ptr<const shared_t>`). These two files arrays.cpp and vectors.cpp were the first I ported and before I wrote that helper; I've moved them over as of ab25667

Reply:
ugh, and then fixed in 71a7d8d